spec/core/ics-026-routing-module/UPGRADES.md

```diff
@@ -0,0 +1,141 @@
+# Application Upgrade Callbacks
+
+## Synopsis
+
+This standard document specifies the interfaces and state machine logic that IBC applications must implement in order to enable existing channels to upgrade their applications after the initial channel handshake.
+
+### Motivation
+
+As new features get added to IBC applications, chains may wish the take advantage of new application features without abandoning the accumulated state and network effect(s) of an already existing channel. The upgrade protocol proposed would allow applications to renegotiate an existing channel to take advantage of new features without having to create a new channel, thus preserving all existing application state while upgradng to new application logic.
+
+
+### Desired Properties
+
+- Both applications MUST agree to the renegotiated application parameters.
+- Application state and logic on both chains SHOULD either be using the old parameters or the new parameters, but MUST NOT be in an in-between state, e.g., it MUST NOT be possible for an application to run v2 logic, while its counterparty is still running v1 logic.
+- The application upgrade protocol is atomic, i.e., 
+  - either it is unsuccessful and then the application MUST fall-back to the original application parameters; 
+  - or it is successful and then both applications MUST adopt the new application parameters and the applications must process packet data appropriately.
+- The application must be able to maintain several different supported versions such that one channel may be on version `v1` and another channel may be on version `v2` and the application can handle the channel state and logic accordingly depending on the application version for the respective channel.
+
+The application upgrade protocol MUST NOT modify the channel identifiers.
+
+## Technical Specification
+
+In order to support channel upgrades, the application must implement the following interface:
+
+```typescript
+interface ModuleUpgradeCallbacks {
+    onChanUpgradeInit: onChanUpgradeInit,
+    onChanUpgradeTry: onChanUpgradeTry,
+    onChanUpgradeAck: onChanUpgradeAck,
+    onChanUpgradeConfirm: onChanUpgradeConfirm,
+    onChanUpgradeRestore: onChanUpgradeRestore
+}
+```
+
+#### **OnChanUpgradeInit**
+
+`onChanUpgradeInit` will verify that the upgrade parameters 
+are valid and perform any custom `UpgradeInit` logic.
+It may return an error if the chosen parameters are invalid 
+in which case the upgrade handshake is aborted.
+If the provided version string is empty, `onChanUpgradeInit` should return 
+a default version string or an error if the provided version is invalid.
```

Comment:
The sentence is unclear.
> or an error if the provided version is invalid.

How can an empty version be invalid?

Also, it seems that the following two sentences are describing the same thing.